raft/src/main/java/org/apache/kafka/raft/utils/VoteRpc.java

```diff
@@ -0,0 +1,259 @@
+                                                                                .directoryId()
+                                                                                .orElse(ReplicaKey.NO_DIRECTORY_ID)
+                                                                )
+                                                                .setLastOffsetEpoch(lastEpoch)
+                                                                .setLastOffset(lastEpochEndOffset)
+                                                )
+                                        )
+                        )
+                );
+    }
+
+    public static VoteResponseData singletonVoteResponse(
+            ListenerName listenerName,
+            short apiVersion,
+            Errors topLevelError,
+            TopicPartition topicPartition,
+            Errors partitionLevelError,
+            int leaderEpoch,
+            int leaderId,
+            boolean voteGranted,
+            Endpoints endpoints
+    ) {
+        VoteResponseData response = new VoteResponseData()
+                .setErrorCode(topLevelError.code())
+                .setTopics(Collections.singletonList(
+                        new VoteResponseData.TopicData()
+                                .setTopicName(topicPartition.topic())
+                                .setPartitions(Collections.singletonList(
+                                        new VoteResponseData.PartitionData()
+                                                .setErrorCode(partitionLevelError.code())
+                                                .setLeaderId(leaderId)
+                                                .setLeaderEpoch(leaderEpoch)
+                                                .setVoteGranted(voteGranted)))));
+
+        if (apiVersion >= 1) {
+            Optional<InetSocketAddress> address = endpoints.address(listenerName);
+            if (address.isPresent() && leaderId >= 0) {
+                // Populate the node endpoints
+                VoteResponseData.NodeEndpointCollection nodeEndpoints = new VoteResponseData.NodeEndpointCollection(1);
+                nodeEndpoints.add(
+                        new VoteResponseData.NodeEndpoint()
+                                .setNodeId(leaderId)
+                                .setHost(address.get().getHostString())
+                                .setPort(address.get().getPort())
+                );
+                response.setNodeEndpoints(nodeEndpoints);
+            }
+        }
+
+        return response;
+    }
+
+    public static Optional<ReplicaKey> voteRequestVoterKey(
+            VoteRequestData request,
+            VoteRequestData.PartitionData partition
+    ) {
+        if (request.voterId() < 0) {
+            return Optional.empty();
+        } else {
+            return Optional.of(ReplicaKey.of(request.voterId(), partition.voterDirectoryId()));
+        }
+    }
+
+    public static AddRaftVoterRequestData addVoterRequest(
```

Comment:
VoteRequest, VoteResponse should be part of the VoteRPCs, but not addVoter, removeVoter, updateVoter. Perhaps we want a DynamicReconfigRpc class for those